src/Bicep.Cli/Services/CompilationService.cs

```diff
@@ -61,11 +63,27 @@ public async Task<Compilation> CompileAsync(string inputPath, bool skipRestore)
                 }
             }
 
-            var compilation = new Compilation(this.invocationContext.ResourceTypeProvider, sourceFileGrouping);
+            try
+            {
+                var compilation = new Compilation(this.invocationContext.ResourceTypeProvider, sourceFileGrouping);
+
+                LogDiagnostics(compilation);
+
+                return compilation;
+            }
+            catch (Exception exception)
```

Comment:
>exception [](http://example.com/codeflow?start=29&length=9)

We shouldn't blindly catch all the exceptions and turn them into diagnostics.

Reply:
Hmmm, we do this in language server as well:
https://github.com/Azure/bicep/blob/d2ee05f6fcc2043f0b15bf04af2b3e97aded653f/src/Bicep.LangServer/BicepCompilationManager.cs#L247

On the language server side, if the config file is invalid, we don't proceed till config file is error free. The current compilation design needs a config file. 

What do we want the behavior to be? If we encounter invalid config file, log diagnostic and use default config settings? That would be confusing. 

One option I can think of:
- If the config file is invalid, log error diagnostic 
- Use disabled config, something similar to what we do in tests  to create compilation https://github.com/Azure/bicep/blob/632fa2d86ef1d7da61029ab49a59de6e27a4b1f9/src/Bicep.Core.UnitTests/Configuration/ConfigHelperExtensions.cs#L16
- Push other diagnostics along with the error from invalid config

Let me know your thoughts. Adding @anthony-c-martin as well.

Reply:
We should definitely surface the exception to the user, but until now, we've just relied on the try-catch in `Program.RunAsync` for this. Is there a reason to write a diagnostic instead of following this pattern?

Reply:
Do you mean use InvocationContext to write error? That would look like this:
![image](https://user-images.githubusercontent.com/30270536/133333724-952c9f12-3306-4d50-9e7d-920ae73f00a2.png)

Reply:
Yeah - in the above screenshot, the exception message itself is definitely useful, but I don't think there's any need to log the stack trace. The only thing I'd change in the message is making sure we log **which** file couldn't be parsed - so the user knows where to look for the bad configuration file.

Reply:
Created below issue to track the issue on language server side:
https://github.com/Azure/bicep/issues/4471

Reply:
Made changes to write error message and go ahead with compilation creation:
https://github.com/Azure/bicep/commit/699af2a087c920c1002078f3d05e2016783c7f07